### DISCUSSION TITLE
IterableWeakMap instead of WeakMap for instrumentation and dev tools?

### DISCUSSION THREAD

**Discussion by loganvolkers** [pmndrs/jotai#2116] [Ideas] [answered]
Internally the `store` in Jotai is based on `WeakMap`. This is a great advantage because atoms are garbage collected automatically.

Unfortunately, this also means that it makes it hard to see what is currently stored in state to build things like devtool plugins. The workaround of this has been to replace `WeakMap` with just `Map` in dev environments. Alternatively, we should consider using an `IterableWeakMap` to allow internal APIs to be connected into dev tools.

See the proposal on WeakRefs for a complete implementation of a IterableWeakMap that uses `FinalizationRegistry` and `WeakRef` https://github.com/tc39/proposal-weakrefs#iterable-weakmaps

Opening this thread as a point of discussion for internal collaborators to consider. I have no experience with `FinalizationRegistry` and am also not experienced with dev tools, but recognize the value that Pinia and Redux bring with their dev tools, and the barrier that `WeakMap` creates.

**Comment by loganvolkers** [comment]
Note the implementation of an `IterableWeakMap` [in TC39](https://github.com/tc39/proposal-weakrefs#iterable-weakmaps) linked above [may have a bug](https://github.com/tc39/proposal-weakrefs/issues/213)

**Comment by dai-shi** [comment, marked as answer]
I haven't heard of IterableWeakMap. Thanks for sharing.
Though, I'm skeptical about `FinalizationRegistry`. I've tried it once in another project of mine, and found it hard and less predictable. If we have a stable `IterableWeakMap`, I'm fine to consider some opt-in option.

Meanwhile, would you try some DEV-only store methods?
https://github.com/pmndrs/jotai/blob/f7d7f74eb64660421dfea78d2be1e47816ffd6be/src/vanilla/store.ts#L750-L776

Those are what `jotai-devtools` uses.

I now think those dev methods are too many, we could just return a `Map`. cc @arjunvegda
